### ISSUE TITLE
ID Embedding optimization loss 

### ISSUE THREAD

**Issue by aldipiroli** [deeplearning-wisc/npos#3]
Hello and thanks for sharing the awesome work! 
I had some questions regarding the ID Embedding optimization described in the paper:

1.  I was looking at the repo and I did not find the [DispLoss and CompLoss](https://github.com/deeplearning-wisc/npos/blob/34fc4440598162b63d1ca8408f9c554b37af40c5/training_from_sctrach/CIFAR/train_CIFAR10.py#L26) used for the ID feature optimization.  I also had a look at the [clip](https://github.com/openai/CLIP) and [faiss](https://github.com/facebookresearch/faiss) but also did not find them there. 
2.  For the loss calculation of R_closed in the [code](https://github.com/deeplearning-wisc/npos/blob/34fc4440598162b63d1ca8408f9c554b37af40c5/training_from_sctrach/CIFAR/train_CIFAR10.py#L326)  you use _features_ `normed_features = F.normalize(features, dim=1)` which is derived from 
  ```
 penultimate = model.encoder(input)
 features = model.head(penultimate)
```
As far as I understand,   `model.head` is not the last classification layer. So I was wandering, if  model.head is the penultimate layer before the classification, why not use _features_ as ID features instead of _penultimate_?

3. Finally, since this seems like a natural improvement over VOS, I was wandering if you ever tested it for object detection? 

Thanks again!

**Comment by taoleitian** [comment]
1. Sorry for the mistake. I have uploaded the loss.py for the [DispLoss and CompLoss](https://github.com/deeplearning-wisc/npos/blob/34fc4440598162b63d1ca8408f9c554b37af40c5/training_from_sctrach/CIFAR/train_CIFAR10.py#L26). This version of the code should be updated in the following days.
2. In fact, the head is the last layer, because we use contrastive learning to aggregate features, we do not explicitly use the fully connected layer for classification, but use the prototype of each category. As stated in the paper, we use the feature of the last layer as the ID feature. Actually, I don't know what you're curious about. If possible, further statements are welcome.
3. Since NPOS needs to adjust far more parameters than VOS, we only conducted experiments on classification instead of detection.
